cpp/src/arrow/ipc/metadata_internal.cc

```diff
@@ -367,8 +367,8 @@ Status ConcreteTypeFromFlatbuffer(flatbuf::Type type, const void* type_data,
         return Status::Invalid("Map's keys must be non-nullable");
       } else {
         auto map = static_cast<const flatbuf::Map*>(type_data);
-        *out = std::make_shared<MapType>(children[0]->type()->field(0)->type(),
-                                         children[0]->type()->field(1)->type(),
+        *out = std::make_shared<MapType>(children[0]->type()->field(0)->WithName("key"),
```

Comment:
How about this? Nullability of key field has been checked above. So here I just enforce key/value field name with other attributes unchanged. @lidavidm @wjones127

Reply:
I think that's OK, since that's effectively what it was doing before when we just took the type. And now we preserve metadata.

Reply:
However we should also probably force the nullability of key to `false` here to further match existing behavior.

Reply:
Line 366 has checked this already. I don't think we need to explicitly set it to false here.

Reply:
Ah, d'oh. thanks